test: Convert to MTEExtension

keturn opened the pull request:
Fixes errors about tests failing to resolve modules.

Also re-enabled some tests and got them working. I changed enough of the test file that GitHub PR interface doesn't recognize it as a file that moved, but it is making the same assertions.

This PR is also notable because it takes care of the **last** use of the deprecated ModuleTestingEnvironment superclass in Omega.

## To Do
- [x] https://github.com/Terasology/ModuleTestingEnvironment/pull/64
- [x] #107

keturn committed:
test: Convert to MTEExtension

keturn committed:
Merge remote-tracking branch 'origin/develop' into test/jupiter-extension

# Conflicts:
#	src/test/java/regions/RegionEntitiesTest.java

keturn committed:
test: update MTE version

keturn committed:
test: lint 🚮

keturn committed:
test(RegionEntitiesTest): re-enable tests

use parameterized test to help clarify in the test runner which things are passing

keturn commented:
Results improve _slightly,_ at the cost of a much _much_ longer runtime, by using IsolatedMTEExtension.

Which I find puzzling in itself, because the code executed in the tests looks like it should all be read-only operations? So why would they benefit from increased isolation?

keturn commented:
Also: The `cellIsLoaded` tests failing for anything that isn't (0, 0) could be explained by it only loading the one thing at the origin, and it returns false for everything else.

Except there is a _failing_ test saying the cell at (98, -124) ***is*** loaded when it _shouldn't_ be?

keturn committed:
chore: lint 🚮

keturn committed:
test: move test to match package of class

keturn commented:
This led to some MTE work for untangling interactions between test cases: https://github.com/Terasology/ModuleTestingEnvironment/pull/64

also: a lot of the code under test here in RegionEntitesTest changed for JOML in #86. If these tests were disabled (or just ignored) at the time, something might have slipped through then.

keturn committed:
test(RegionEntitiesTest): use PER_CLASS lifecycle so we can run setupEntityRefs just once

skaldarnar commented on src/main/java/org/terasology/dynamicCities/region/RegionEntityManager.java:
```diff
@@ -75,18 +72,18 @@ public void postBegin() {
         regionEntitiesComponent = new RegionEntitiesComponent(gridSize);
         regionStoreEntity = entityManager.create(regionEntitiesComponent, new RegionMainStoreComponent());
-        regionStoreEntity.setAlwaysRelevant(true);
+        regionStoreEntity.setScope(EntityScope.GLOBAL);
```
I think @DarkWeird mentioned something about removing sectors (and therefore scopes?). Just to make people aware ...

keturn replied:
This change was because `setAlwaysRelevant` is deprecated, and `setScope` is what it pointed me to.

DarkWeird replied:
Global scope exists yet.

keturn commented:
![](https://github.trello.services/images/mini-trello-icon.png) [MTE tests: Dynamic Cities](https://trello.com/c/ugRH2AfH/176-mte-tests-dynamic-cities)

keturn commented:
![](https://github.trello.services/images/mini-trello-icon.png) [MTE tests: Dynamic Cities](https://trello.com/c/ugRH2AfH/176-mte-tests-dynamic-cities)

keturn committed:
fix(RegionEntityManager.getNearest): removing a conspicuous -1 makes it fit the tests

Last changed here: https://github.com/Terasology/DynamicCities/pull/81/files#diff-d87e51cd5a25980fad8f20ba460f2f8824e5d7f2a50cbd53c278babe35cea074

keturn committed:
chore: remove obsolete debugging comments

keturn committed:
test: comment on remaining disabled test

keturn commented on src/main/java/org/terasology/dynamicCities/region/RegionEntityManager.java:
```diff
@@ -123,8 +120,8 @@ public EntityRef get(Vector2ic position) {
 
     public EntityRef getNearest(Vector2ic position) {
         Map<Vector2i, EntityRef> regionEntities = regionEntitiesComponent.regionEntities;
-        int x = Chunks.toChunkPosX(position.x()) * Chunks.SIZE_X + ((Chunks.SIZE_X / 2) - 1);
-        int y = Chunks.toChunkPosZ(position.y()) * Chunks.SIZE_Z + ((Chunks.SIZE_Z / 2) - 1);
+        int x = Chunks.toChunkPosX(position.x()) * Chunks.SIZE_X + Chunks.SIZE_X / 2;
+        int y = Chunks.toChunkPosZ(position.y()) * Chunks.SIZE_Z + Chunks.SIZE_Z / 2;
```
:warning: This is a real change, not just punctuation for CheckStyle.

This makes the results line up with what the tests have always expected, contrary to the results produced by #81.

jdrueckert replied:
Have you checked in MetalRenegades that cities and buildings still show up?

jdrueckert replied:
Have you checked in MetalRenegades that cities and buildings still show up?

keturn replied:
I've reverted this change, as it makes the placeParcel test fail.

jdrueckert commented on src/main/java/org/terasology/dynamicCities/region/RegionEntityManager.java:
```diff
@@ -273,10 +272,8 @@ public void setNameTagForRegion(EntityRef region) {
     }
 
-    @Command(shortDescription = "Toggles the view of nametags for region entities", runOnServer = true,
-            requiredPermission = PermissionManager.DEBUG_PERMISSION)
```
Is this permission not required after all? Or why was it removed?

keturn replied:
Code inspection told me that `DEBUG_PERMISSION` is the default and suggested omitting it.

jdrueckert reviewed (dismissed):
@keturn I'm approving this assuming that you checked in MetalRenegades, that cities and buildings still show up with your change in the `RegionEntitiesManager`. Please only merge if you did ^^

keturn commented:
> I'm approving this assuming that you checked in MetalRenegades, that cities and buildings still show up […]

Cities? We know, after several months of playtesting, that Cities are a myth. Propaganda by the liberal elite to destabilize the traditional way of life by putting stories in the minds of our children, fantasies about the great named _Cities_ and their _buildings_ taller than trees. Hah!

---

Oh, you mean in single-player, I should still expect to see cities and buildings in single-player?

Well, I have this much:  
![1](https://user-images.githubusercontent.com/83819/133495115-41b6bbbb-b772-4adc-8299-5d11476acdb4.jpg)

Is that a city?

As for buildings, well,

> 12:02:12.454 [main] ERROR o.t.engine.core.TerasologyEngine - Uncaught exception, attempting clean game shutdown  
> java.lang.NullPointerException: null  
	at org.terasology.dynamicCities.settlements.SettlementEntityManager.growSettlement(SettlementEntityManager.java:531)  
>	at org.terasology.dynamicCities.settlements.SettlementEntityManager.onWorldTimeEvent(SettlementEntityManager.java:168)

https://github.com/Terasology/DynamicCities/blob/3a609aa3b0c3e08d1529e52211ad22afc2d91887/src/main/java/org/terasology/dynamicCities/settlements/SettlementEntityManager.java#L522-L535

keturn commented:
Second attempt: didn't get _that_ crash, but didn't get buildings.

Tried on `develop` for comparison: also has Problems. See [thread](https://discord.com/channels/270264625419911192/887780121535201290/887804870332145664 "#SettlementEntityManager crash").

pollend commented:
> Second attempt: didn't get _that_ crash, but didn't get buildings.
> 
> Tried on `develop` for comparison: also has Problems. See [thread](https://discord.com/channels/270264625419911192/887780121535201290/887804870332145664).

the code is way to fragile ;/ might have something to do with all the terrain features.

Cervator commented:
Can confirm an apparent and distinct lack of buildings - darn termites!

Noticed some logging I don't recall seeing before about broken parcels?

```
15:35:54.794 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:36:00.089 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1094), capping to 1000
15:36:46.281 [main] ERROR o.t.e.e.e.internal.PojoEntityManager - Adding a component (class org.terasology.engine.network.NetworkComponent) over an existing component for entity 23300
15:36:46.299 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4cce9387 couldn't be completed. Status: NONE
15:36:46.309 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7be79549 couldn't be completed. Status: NONE
15:36:53.422 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:36:53.423 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:36:54.878 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:37:01.166 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1004), capping to 1000
15:37:05.165 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:37:05.167 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:37:07.380 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:37:21.675 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:37:21.677 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:37:22.151 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@649b9ed couldn't be completed. Status: NONE
15:37:22.151 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4cce9387 couldn't be completed. Status: NONE
15:37:22.162 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7ebbda4f couldn't be completed. Status: NONE
15:37:22.162 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7be79549 couldn't be completed. Status: NONE
15:37:24.184 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:37:27.554 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:37:27.555 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:37:30.346 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:37:32.961 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:37:32.962 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:37:33.787 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:37:37.137 [main] WARN  o.terasology.biomesAPI.BiomeManager - Missing biome information for ( 3.510E+2  2.100E+1  1.183E+3)
15:37:44.864 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1903), capping to 1000
15:37:49.489 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1069), capping to 1000
15:37:59.223 [main] ERROR o.t.e.e.e.internal.PojoEntityManager - Adding a component (class org.terasology.engine.network.NetworkComponent) over an existing component for entity 42586
15:37:59.231 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@649b9ed couldn't be completed. Status: NONE
15:37:59.232 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4cce9387 couldn't be completed. Status: NONE
15:37:59.232 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@1bc6d1ba couldn't be completed. Status: NONE
15:37:59.246 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@2bcf9031 couldn't be completed. Status: NONE
15:37:59.257 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7ebbda4f couldn't be completed. Status: NONE
15:37:59.257 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@6666978e couldn't be completed. Status: NONE
15:37:59.257 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7be79549 couldn't be completed. Status: NONE
15:38:00.033 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:38:00.034 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:38:04.709 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:38:15.173 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1523), capping to 1000
15:38:21.449 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:38:21.451 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:38:21.761 [main] INFO  o.t.engine.logic.console.ConsoleImpl - [CONSOLE] Player86648 (local) entered Kinhestcasou
15:38:24.690 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
15:38:26.112 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1152), capping to 1000
15:38:32.781 [main] INFO  o.t.engine.logic.console.ConsoleImpl - [CONSOLE] Player86648 (local) left Kinhestcasou
15:38:33.899 [main] INFO  o.t.engine.logic.console.ConsoleImpl - [CONSOLE] Player86648 (local) entered Kinhestcasou
15:38:35.816 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@649b9ed couldn't be completed. Status: NONE
15:38:35.817 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4cce9387 couldn't be completed. Status: NONE
15:38:35.817 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@643f2816 couldn't be completed. Status: NONE
15:38:35.817 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@1bc6d1ba couldn't be completed. Status: NONE
15:38:35.838 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4b313414 couldn't be completed. Status: NONE
15:38:35.838 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@2bcf9031 couldn't be completed. Status: NONE
15:38:35.850 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@484f898e couldn't be completed. Status: NONE
15:38:35.850 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7ebbda4f couldn't be completed. Status: NONE
15:38:35.850 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@6666978e couldn't be completed. Status: NONE
15:38:35.850 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7be79549 couldn't be completed. Status: NONE
15:38:42.151 [main] WARN  o.t.engine.core.internal.TimeBase - Delta too great (1010), capping to 1000
15:38:52.640 [o.t.e.r.n.l.m.a.AdvancedGameSetupScreen-0] INFO  o.t.e.c.module.ModuleListDownloader - Retrieved 185 entries
15:38:57.322 [main] INFO  o.t.engine.logic.console.ConsoleImpl - [CONSOLE] Ghost mode toggled
15:39:11.796 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@649b9ed couldn't be completed. Status: NONE
15:39:11.797 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@2031e024 couldn't be completed. Status: NONE
15:39:11.797 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4cce9387 couldn't be completed. Status: NONE
15:39:11.797 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@643f2816 couldn't be completed. Status: NONE
15:39:11.797 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@1bc6d1ba couldn't be completed. Status: NONE
15:39:11.806 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@22d7be07 couldn't be completed. Status: NONE
15:39:11.806 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@4b313414 couldn't be completed. Status: NONE
15:39:11.806 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@2bcf9031 couldn't be completed. Status: NONE
15:39:11.813 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@484f898e couldn't be completed. Status: NONE
15:39:11.813 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7ebbda4f couldn't be completed. Status: NONE
15:39:11.813 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@704f88f8 couldn't be completed. Status: NONE
15:39:11.813 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@6666978e couldn't be completed. Status: NONE
15:39:11.813 [main] WARN  o.t.d.s.SettlementEntityManager - Parcel org.terasology.dynamicCities.parcels.RoadParcel@7be79549 couldn't be completed. Status: NONE
15:39:21.460 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
15:39:21.461 [main] INFO  o.t.e.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
15:39:23.477 [Saving-0] INFO  o.t.e.p.internal.SaveTransaction - Save game finished
```

Checked two city spots, then also 0, 0, 0 as some past bugs have sent buildings back there, but nope, nothing. So maybe something is keep parcels from forming correctly? Otherwise the new MR worlds are gorgeous :-)

keturn commented:
Example to compare to: [Earlier](Niruandaleth) used seed ​`iTKAXPCM7KkMFnBj` and found a city at (-530, 12, -70) that looked like this:

![At least half a dozen buildings of several kinds, widely spaced.](https://cdn.discordapp.com/attachments/580146738967347211/891421906333478962/unknown.png)

keturn committed:
Merge branch 'develop' into test/jupiter-extension

# Conflicts:
#	module.txt
#	src/main/java/org/terasology/dynamicCities/region/RegionEntityManager.java

keturn commented:
Okay, now that #107 is merged (thank you Niruandaleth for the review) we have a properly failing test in this branch!

and I've confirmed that if I re-add the `-1`s in RegionEntityManager, it makes some warnings that come up during the `placeParcel` test go away. Worryingly, it doesn't make it _pass._

I still want to track down where the setter for that is and figure out an interface that lets us avoid re-doing the math in different places.

keturn commented:
The location that `getNearest` is trying to match is set by RegionEntityProvider: https://github.com/Terasology/DynamicCities/blob/8bd4203393a9860ea766816a985eaa3fe24eaf82/src/main/java/org/terasology/dynamicCities/region/RegionEntityProvider.java#L64-L65
(It makes entities with an `UnregisteredRegionComponent`, and the RegionEntityManager system has an OnActivatedComponent handler that receives those and invokes `add()` with them.)

My notes so far:
* `getNearest` should _always_ succeed for any coordinates (x, z) for which the world has generated a chunk at the _surface_ for that location.
  * That is, it should only fail if, say, there is a very very tall cliff, and when we stand at the top of the cliff, the surface below is so low that it hasn't even been close enough to us to generate yet.
  * _Every_ location on the surface should map to _some_ region entity.
* The location is set as the center of the Region given to the EntityProvider.
* `getNearest` tries to calculate a location to align with that, assuming that regions are always aligned along chunk boundaries. That _could_ be true, but the assumption violates some abstraction and it's fragile if the world generator ever changes some decisions about regions.
* I still haven't figured out what _did_ change. Seems like the RegionEntityProvider code has always used `center()` for that Location. Did the pre-JOML center method round differently? Or is it something else about regions that changed, like their boundary size?

keturn commented:
Next steps?

* Does the engine offer any better way to store data aggregated by chunk?
  * (or, perhaps in this case, the column of chunks at (x, z)?)
* If not and we do continue with the current implementation, is there a better way to normalize these coordinates?
  * Is "Region" a concept we expect to extend to non-World-Generation game systems at all?
  * Is there any guarantee that a Region contains _only one_ Chunk that is unique to it?

keturn commented:
We had a discussion about Dynamic Cities in today's meeting, and it went sprawling as these things do, but I think the the parts that are relevant for next steps here are:

- We don't want to drop Metal Renegades. So Dynamic Cities is sticking around. This did spark some discussion about whether Static Cities would replace Dynamic Cities in the Metal Renegades lineup, whether that would actually be more maintainable, whether it would need additional features to get there, etc, but that's all To Be Determined.

- We don't want our confusion around Dynamic Cities to block the original goal of this PR: completing the removal of the deprecated ModuleTestingEnvironment superclass. So we should `@Disable` or kludge things as necessary here to get this code to a place where the tests that are enabled aren't misleading, and manual testing shows no regressions (i.e. cities have buildings in single-player mode).

keturn committed:
fix(RegionEntityManager): revert change to getNearest

adjust tests to compensate for the mysterious -1

keturn commented:
I've set aside my quest to track down the origin of the `-1` and reverted my change to `getNearest` so that the `placeParcel` test passes again.

I started a new MR world with the [known seed](https://github.com/Terasology/DynamicCities/pull/94#issuecomment-938183862) and did find a city with two buildings there.

I'd offer a screenshot, but the screenshot pushed the RAM usage over the edge and it crashed.

jdrueckert commented:
Have a screenshot :palms_up_together:
![Terasology-211102195103-2560x1395](https://user-images.githubusercontent.com/29981695/139927829-7889901d-5a02-4eb6-971a-7bf7e229074b.png)
